File: breadth-first-search/word-ladder/word-ladder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
function ladderLength(beginWord, endWord, wordList) {
  const set = new Set(wordList);

  if (!set.has(endWord)) {
    return 0;
  }

  const queue = [];

  queue.push(beginWord);

  let level = 1;

  while (queue.length) {
    const size = queue.length;

    for (let i = 0; i < size; i++) {
      const curr = queue.shift();
      const chars = curr.split('');

      for (let j = 0; j < chars.length; j++) {
        const original = chars[j];
        for (
          let c = 'a';
          c.charCodeAt(0) <= 'z'.charCodeAt(0);
          c = String.fromCharCode(c.charCodeAt(0) + 1)
        ) {
          if (chars[j] == c) {
            continue;
          }

          chars[j] = c;

          const newWord = chars.join('');

          if (newWord === endWord) {
            return level + 1;
          }

          if (set.has(newWord)) {
            queue.push(newWord);
            set.delete(newWord);
          }
        }
        chars[j] = original;
      }
    }
    level++;
  }
  return 0;
}

File: breadth-first-search/word-ladder/word-ladder.go
func ladderLength(beginWord string, endWord string, wordList []string) int {
	set := make(map[string]bool)

	for _, word := range wordList {
		set[word] = true
	}

	if _, ok := set[endWord]; !ok {
		return 0
	}

	queue := []string{}

	level := 1

	queue = append(queue, beginWord)

	for len(queue) > 0 {
		size := len(queue)

		for i := 0; i < size; i++ {
			chars := []byte(queue[0])
			queue = queue[1:]

			for j := 0; j < len(chars); j++ {
				original := chars[j]

				for c := byte('a'); c <= byte('z'); c++ {
					if original == c {
						continue
					}

					chars[j] = c

					newWord := string(chars)

					if newWord == endWord {
						return level + 1
					}

					if _, ok := set[newWord]; ok {
						queue = append(queue, newWord)
						delete(set, newWord)
					}
				}
				chars[j] = original
			}
		}
		level++
	}
	return 0
}
